release-2.1: storage: don't clear lastTxnMeta on WriteIntentError to different key

nvb opened the pull request:
Backport 1/1 commits from #32773.

/cc @cockroachdb/release

---

Fixes #32582.

This change removes a faulty optimization in the `contentionQueue`.
The optimization removed the txnMeta associated with a contended key
in the queue when it found a `WriteIntentError` from a different key.
It didn't take into account that this error could be from an earlier
request within the same batch, meaning that we can't make any assumptions
about the state of the previously contended intent simply because we
see a different `WriteIntentError`.

Release note (bug fix): Fix a bug where metadata about contended keys
was inadvertently ignored, allowing for a failure in txn cycle detection
and transaction deadlocks in rare cases.

nvb committed:
storage: don't clear lastTxnMeta on WriteIntentError to different key

Fixes #32582.

This change removes a faulty optimization in the `contentionQueue`.
The optimization removed the txnMeta associated with a contended key
in the queue when it found a `WriteIntentError` from a different key.
It didn't take into account that this error could be from an earlier
request within the same batch, meaning that we can't make any assumptions
about the state of the previously contended intent simply because we
see a different `WriteIntentError`.

Release note (bug fix): Fix a bug where metadata about contended keys
was inadvertently ignored, allowing for a failure in txn cycle detection
and transaction deadlocks in rare cases.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/32853)
<!-- Reviewable:end -->

nvb committed:
storage: fix flake in TestContendedIntentChangesOnRetry

Saw in CI of #32853.

We were ignoring deadlock errors on txn4, but not txn5.

Release note: None

nvanbenschoten commented:
Waiting on https://github.com/cockroachdb/cockroach/pull/32864.

nvanbenschoten commented:
Cherry-picked #32864. Merging on green.
